### PR TITLE
kubevirt: upgrade k3s,multus,kubevirt,cdi,longhorn

### PR DESCRIPTION
**Pull request by andrewd-zededa**
component upgrades, info messages
A kube container generation number is tracked in cluster-update.sh and compared to an 'applied' version in the persistent node fs /var/lib.
When the applied version is behind it triggers component upgrades.

cluster-update.sh uses an inline command in the zedkube micro service to publish status updates to zedagent which will send info messages (ZInfoKubeClusterUpdateStatus) to a controller.

### CODE REVIEW

**Comment by andrewd-zededa** [comment]
The vendor directory really increased the commit size, otherwise its significantly smaller in file count.

**Comment by andrewd-zededa** [comment]
I'll rebase off latest poc-aug13-2024 and then get it ready for review

**Comment by andrewd-zededa** [comment]
I noticed the diag messages showing drain progress stopped working, I'm trying to resolve that.

**Comment by andrewd-zededa** [comment]
The node drain request in the baseos path might need to be moved to the activate step to handle downgrades where the baseos is already written.

**Comment by andrewd-zededa** [comment]
Latest push includes the following:
- baseosmgr no longer requests drain before baseos activation.
- ZInfoKubeClusterUpdateStatus info messages are sent on every pubsub create/modify of KubeClusterUpdateStatus
- diag node drain prints include a timestamp

I'll rebase off latest next.

File: pkg/pillar/kubeapi/nodedrain.go [hunk 1]
```diff
@@ -35,20 +35,32 @@ func RequestNodeDrain(pubNodeDrainRequest pubsub.Publication, requester DrainReq
...
 				return &cfg
 			}
 		}
 	}
 	return nil
 }
 
+// DrainStatusFaultInjection_Wait : while this file exists, wait in the drain status
+func DrainStatus_FaultInjection_Wait() bool {
+	injectFaultPath := "/tmp/DrainStatus_FaultInjection_Wait"
```
Comment (zedi-pramodh) [inline pkg/pillar/kubeapi/nodedrain.go]: Is this a temp file and used only in runtime ?

Comment (andrewd-zededa) [reply pkg/pillar/kubeapi/nodedrain.go]: yeah, would only exist for that boot, next boot would have a clean overlays.

File: pkg/pillar/docs/zedkube.md [hunk 2]
```diff
@@ -107,8 +106,8 @@ NodeDrainStatus can be forced by writing the object (in pillar svc container fs)
 eg. to force disable drain:
 echo '{"Status":1,"RequestedBy":1}' > /tmp/force-NodeDrainStatus-global.json
 
-eg. to force drain complete:
-echo '{"Status":1,"RequestedBy":1}' > /tmp/force-NodeDrainStatus-global.json
+eg. to force deviceop drain complete:
+echo '{"Status":9,"RequestedBy":2}' > /tmp/force-NodeDrainStatus-global.json
```
Comment (zedi-pramodh) [inline pkg/pillar/docs/zedkube.md]: In other file you changed this path to /persist ? You need to update in this file.

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 3]
```diff
@@ -2035,6 +2069,50 @@ func triggerPublishHwInfo(ctxPtr *zedagentContext) {
...
+}
+func handleClusterUpdateStatusModify(ctxArg interface{}, key string,
+	configArg interface{}, oldStatusArg interface{}) {
+	handleClusterUpdateStatusImpl(ctxArg, key, configArg, oldStatusArg)
+}
+func handleClusterUpdateStatusImpl(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+	ctx, ok := ctxArg.(*zedagentContext)
+	if !ok {
+		log.Fatalf("handleClusterUpdateStatusImpl invalid type in ctxArg: %v", ctxArg)
```
Comment (zedi-pramodh) [inline pkg/pillar/cmd/zedagent/zedagent.go]: Do we really have to log fatal ? Remember fatal will reboot the device.  FYI, Rebooting the device during upgrade may not work out well. I will leave it to you.

Comment (zedi-pramodh) [reply pkg/pillar/cmd/zedagent/zedagent.go]: Also we cannot fatal just for invalid args

**Review by zedi-pramodh — APPROVED** [review]
Too much code. I will say ship it and test it. We will review anyway when merging to master.

**Pull request merged**